util/mqtt.c

```diff
@@ -491,3 +489,235 @@ mqtt_publish_single_message (const char *server_uri_in, const char *topic,
 
   return ret;
 }
+
+/**
+ * @brief subscribes to a single topic.
+ *
+ * mqtt_subscribe_r uses given mqtt_t to subscribe with given qos to given
+ * topic.
+ *
+ * To be able to subscribe to a topic the client needs to be connected to a
+ * broker.
+ *
+ * @param mqtt_t	contains the mqtt client
+ * @param qos	quality of service of messages within topic
+ * @param topic         Topic to subscribe to
+ *
+ * @return 0 on success, -1 when given mqtt is not useable, -2 when subscription
+ * failed.
+ */
+int
+mqtt_subscribe_r (mqtt_t *mqtt, int qos, const char *topic)
+{
+  int rc = 0;
+  if (mqtt == NULL || mqtt->client == NULL)
+    {
+      rc = -1;
+      goto exit;
+    }
+  MQTTSubscribe_options opts = MQTTSubscribe_options_initializer;
+  MQTTProperties props = MQTTProperties_initializer;
+  MQTTResponse resp =
+    MQTTClient_subscribe5 (mqtt->client, topic, qos, &opts, &props);
+  if (resp.reasonCode != MQTTREASONCODE_GRANTED_QOS_1)
+    {
+      rc = -2;
+    }
+exit:
```

Comment:
```suggestion
```
You only use the `goto` once ... I am no fan of `goto` because it jumps in the code and makes it unreadable ... also it is only used once, so I do not see any advantage of it here ...